relayer/tools/gen-mmr.go

```diff
@@ -0,0 +1,176 @@
+package main
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"fmt"
+	"hash"
+
+	"reflect"
+
+	"github.com/cbergoon/merkletree"
+	secp256k1 "github.com/ethereum/go-ethereum/crypto"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/sha3"
+)
+
+// Validator keypair (secp256k1)
+type Keypair struct {
+	publicBytes []byte
+	public      *ecdsa.PublicKey
+	private     *ecdsa.PrivateKey
+}
+
+type ValidatorPublicKey []byte
+type ParaBlockHash [32]byte
+
+// Uninterpreted hash type
+type H256 [32]byte
+
+// Leaf node for MMR
+type MMRLeaf struct {
+	// Relay chain block hash
+	blockHash H256
+	// Merkle root for Parachain block header hashes
+	paraHeadsRoot H256
+	// Has validator merkle roor
+	hasValidatorRoot bool
+	// Merkle root for validator public keys
+	validatorRoot H256
+}
```

Comment:
It seems to me that we are mixing up a MMR leaf and a commitment here. Do you have a link to the data structure of the MMR leafs? I was unable to find it in https://github.com/paritytech/grandpa-bridge-gadget. 

The commitment should contain the MMR root and have the following data structure: 

```rust
/// A commitment signed by GRANDPA validators as part of BEEFY protocol.
///
/// The commitment contains a [payload] extracted from the finalized block at height [block_number].
/// GRANDPA validators collect signatures on commitments and a stream of such signed commitments
/// (see [SignedCommitment]) forms the BEEFY protocol.
#[derive(Debug, PartialEq, Eq, codec::Encode, codec::Decode)]
pub struct Commitment<TBlockNumber, TPayload> {
	/// The payload being signed.
	///
	/// This should be some form of cumulative representation of the chain (think MMR root hash).
	/// For transition blocks it also MUST contain details of the next validator set.
	pub payload: TPayload,

	/// Finalized block number this commitment is for.
	///
	/// GRANDPA validators agree on a block they create a commitment for and start collecting
	/// signatures. This process is called a round.
	/// There might be multiple rounds in progress (depending on the block choice rule), however
	/// since the payload is supposed to be cumulative, it is not required to import all
	/// commitments.
	/// BEEFY light client is expected to import at least one commitment per epoch (the one with
	/// [is_set_transition_block] set), but is free to import as many as it requires.
	pub block_number: TBlockNumber,

	/// BEEFY validator set supposed to sign this commitment.
	///
	/// Validator set is changing once per epoch in the commitment with [is_set_transition_block]
	/// set to `true`. Such "epoch commitments" MUST provide the light client with details of the
	/// new validator set as part of the payload. The protocol itself doesn't enforce how these
	/// details are provided though.
	pub validator_set_id: ValidatorSetId,

	/// Indicator of the last block of the epoch.
	///
	/// The payload will contain some form of the NEW validator set public keys information,
	/// yet the block is signed by the current validator set.
	/// When this commitment is imported, the client MUST increment the `validator_set_id`.
	pub is_set_transition_block: bool,
}
```

(From https://github.com/paritytech/grandpa-bridge-gadget/blob/master/primitives/src/commitment.rs#L26)

Reply:
I was using this as a rough reference: https://github.com/Snowfork/verification-solidity/blob/master/contracts/MMRVerification.sol#L8.